### DISCUSSION TITLE
Replacing projection types

### DISCUSSION THREAD

**Discussion by bblfish** [scala/scala3#12527] [General Discussion]
Projection types were removed from Dotty. Some libraries like [banana-rdf](https://github.com/banana-rdf/banana-rdf) relied on this very heavily to make elegant user libraries.

It looks like some subset of it may not be problematic, as per [issue 14: Make type projections safe instead of dropping them](https://github.com/lampepfl/dotty-feature-requests/issues/14).
So we are now in a grey zone where it is not clear whether a safe subset is going to be brought back, or if there is a better way to get the same effect.  So I think we should really have this as a whole discussion topic in itself. Questions to look at:
  - what are the ways to replace projection types, for the various use cases where they have been applied to?
  - how can Scala3 best interact with scala 2.13 libs that use projection types?

**Comment by bblfish** [comment]
Here is a simple interaction problem I have come across regarding inherited projection types.

Banana RDF projection types are defined in the [RDF trait](https://github.com/banana-rdf/banana-rdf/blob/9aa7c4c6daece5713dcafa022ec84e071e9f062c/rdf/shared/src/main/scala/org/w3/banana/RDF.scala#L8) which starts like this:
```scala
trait RDF {
  // types related to the RDF datamodel
  type Graph
  type Triple
  type Node
  type URI <: Node
  type BNode <: Node
  type Literal <: Node
```
This trait is implemented for a number of RDF frameworks such as the Apache Jena project (started at HP in the early 2000s). The [Jena trait](https://github.com/banana-rdf/banana-rdf/blob/9aa7c4c6daece5713dcafa022ec84e071e9f062c/jena/src/main/scala/org/w3/banana/jena/Jena.scala#L15) for example just specifies how those types should be translated for the Jena class hierarchy. 

It looks like implicit transformations which work in Java 2.13 don't work that well anymore when using the library in Scala 3.0. 

We use this for example for our DSL, to allow people to write [RDF in a Scala friendly way](https://github.com/banana-rdf/banana-rdf/blob/742ef2f7adcabbcb4f5334b5788c89a3862a1c8b/rdf-test-suite/shared/src/main/scala/org/w3/banana/diesel/DieselGraphConstructTest.scala#L35):

```Scala
   val g: PointedGraph[Rdf] = (
      bnode("betehess")
      -- foaf.name ->- "Alexandre".lang("fr")
      -- foaf.knows ->- (
        URI("http://bblfish.net/#hjs")
        -- foaf.name ->- "Henry Story"
        -- foaf.currentProject ->- URI("http://webid.info/")
      )
    )
```
This requires Scala to convert a URI into a `PointedGraph` which relies on the [ToPG](https://github.com/banana-rdf/banana-rdf/blob/5a94637abb3fd0bb7cf65fbc9eea7517fe5f4329/rdf/shared/src/main/scala/org/w3/banana/binder/ToPG.scala#L8) Binder. In particular we have 

```scala
 implicit def ToNodeToPG[Rdf <: RDF, T](implicit ops: RDFOps[Rdf], to: ToNode[Rdf, T]): ToPG[Rdf,T] = new ToPG[Rdf, T] {
    def toPG(t: T): PointedGraph[Rdf] = PointedGraph(to.toNode(t))
  }
```

That seems to work fine on Scala 2.13. But for Scala 3 I seem to need to add:

```scala
implicit def URIToNode: ToNode[Rdf,Rdf#URI] = new ToNode[Rdf, Rdf#URI] {
		def toNode(t: Rdf#URI): Rdf#Node = t
	}
```

Is that a bug in Scala3, do I need to write these implicit transformations out explicitly for the inheritance hierarchy, or is there a better way to get this done?

**Reply by julienrf** [reply]
Hello @bblfish 

> We use this for example for our DSL, to allow people to write [RDF in a Scala friendly way](https://github.com/banana-rdf/banana-rdf/blob/742ef2f7adcabbcb4f5334b5788c89a3862a1c8b/rdf-test-suite/shared/src/main/scala/org/w3/banana/diesel/DieselGraphConstructTest.scala#L35):

Could you please elaborate on why you need to use type projections to provide this user-facing DSL?

**Reply by bblfish** [reply]
The main reason for type projections is that it allows one to write code that does not depend on the specific RDF library used. It is then possible to change implementation by changing one line of code. 

All our tests are written this way. For example [GraphUnionTest](https://github.com/banana-rdf/banana-rdf/blob/9aa7c4c6daece5713dcafa022ec84e071e9f062c/rdf-test-suite/shared/src/main/scala/org/w3/banana/GraphUnionTest.scala)
```Scala
abstract class GraphUnionTest[Rdf <: RDF](implicit ops: RDFOps[Rdf]) extends WordSpec with Matchers {
...
"union of Nil must return an empty graph" in {    
     val result: Rdf#Graph = union(Nil)    isomorphism(result, emptyGraph) shouldEqual true  }
...
```

We can then implement these tests by setting the RDF to be Apache's Jena for example as in [JenaTest.scala](https://github.com/banana-rdf/banana-rdf/blob/series/0.8.x/jena/src/test/scala/org/w3/banana/jena/JenaTest.scala) 

```scala
class JenaGraphUnionTest extends GraphUnionTest[Jena]
```
or to IBM's  [RDF4J.scala](https://github.com/banana-rdf/banana-rdf/blob/series/0.8.x/rdf4j/src/test/scala/org/w3/banana/rdf4j/Rdf4jTest.scala). 

In fact you can write your whole codebase generically and switch between implementations by changing a couple of lines of code. This is very useful as it allows one to write code for the client in ScalaJS and on the server, and also try out how well different inferencing engines are working without needing to rewrite one's whole codebase. You could also have your code use both and exchange graphs between the two implementations if needed.

**Reply by julienrf** [reply]
Would it be possible to write something like this instead?

~~~ scala
trait GraphUnionTest {
  val rdf: RDF
  implicit def ops: RDFOps[rdf.type]
  "union of Nil must …" in {
    val result: rdf.Graph = ...
  }
}

// and
class JenaGraphUnionTest extends GraphUnionTest {
  lazy val rdf: Jena = ...
  implicit lazy val ops = ...
}
~~~

**Reply by bblfish** [reply]
I started looking at using dependent types 9 months ago in [banana-play](https://github.com/bblfish/banana-play) and opened [a "feature request" 133](https://github.com/lampepfl/dotty-feature-requests/issues/133). It did not seem that easy to get it written out nicely at the time.(But that was after a few years of not having programmed in Scala.) 

Now I have 6 months experience with Scala3 so it will soon be time to look at it again. But it will be a very big change with over 500 errors to be fixed. So I just thought I'd see with others what their experience was.

**Reply by bblfish** [reply]
It could be that Opaque types are useful.
 https://github.com/scala/docs.scala-lang/issues/2158

**Comment by bblfish** [comment]
It looks like Match Types are the answer to my problem. See https://github.com/lampepfl/dotty/issues/13416 and [test code](https://github.com/bblfish/banana-play/blob/master/src/test/scala/ideas/MatchTypes.scala).

**Reply by bblfish** [reply]
Note: with the latest Scala3 it is [possible now to mix opaque and inline](https://github.com/lampepfl/dotty/pull/12815).

**Reply by bblfish** [reply]
Using Opaque types and pattern matching can be done it seems, see https://contributors.scala-lang.org/t/algebras-of-opaque-types-and-pattern-matching/5245

**Reply by bblfish** [reply]
@sciss described how he went about answering his problem [in lampepfl/dotty-feature-request issue 133](https://github.com/lampepfl/dotty-feature-requests/issues/133#issuecomment-910235296).

**Reply by bblfish** [reply]
Match types do seem to work. See [banana-play's RDF.scala](https://github.com/bblfish/banana-play/blob/main/rdf/shared/src/main/scala/org/w3/banana/RDF.scala).

I am able to write code like
```scala
class PG[Rdf <: RDFObj](val pointer: Node[Rdf], val graph: Graph[Rdf])

object PG:
   def apply[Rdf <: RDFObj](node: Node[Rdf])(using ops: RDFOps[Rdf]): PG[Rdf] = 
        new PG[Rdf](node, ops.empty)
```
